File: src/main/java/macro/library/book/Book.java
package macro.library.book;

import macro.library.Util;
import macro.library.author.Author;
import macro.library.database.BookAuthorTable;
import macro.library.database.BookTable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;

/**
 * Created by Macro303 on 2019-Oct-22
 */
public class Book implements Comparable<Book> {
	@NotNull
	private static final Comparator<Book> comparator = Comparator
			.comparing(Book::getTitle)
			.thenComparing(Book::getSubtitle, Util.nullSafeComparator)
			.thenComparing(Book::getFormat)
			.thenComparing(Book::getISBN);
	@NotNull
	private final Isbn isbn;
	@NotNull
	private String title;
	@Nullable
	private String subtitle;
	@NotNull
	private String publisher;
	@NotNull
	private Format format;

	public Book(@NotNull Isbn isbn, @NotNull String title, @Nullable String subtitle, @NotNull String publisher) {
		this(isbn, title, subtitle, publisher, Format.PAPERBACK);
	}

	public Book(@NotNull Isbn isbn, @NotNull String title, @Nullable String subtitle, @NotNull String publisher, @NotNull Format format) {
		this.isbn = isbn;
		this.title = title;
		this.subtitle = subtitle;
		this.publisher = publisher;
		this.format = format;
	}

	public Book add() {
		BookTable.INSTANCE.insert(this);
		return this;
	}

	public Book push() {
		BookTable.INSTANCE.update(this);
		return this;
	}

	public void remove() {
		BookTable.INSTANCE.delete(isbn);
	}

	@Override
	public int compareTo(@NotNull Book other) {
		return comparator.compare(this, other);
	}

	//<editor-fold desc="Getters and Setters" defaultstate="collapsed">
	@NotNull
	public Isbn getISBN() {
		return isbn;
	}

	@NotNull
	public String getTitle() {
		return title;
	}

	public void setTitle(@NotNull String title) {
		this.title = title;
	}

	@Nullable
	public String getSubtitle() {
		return subtitle;
	}

	public void setSubtitle(@Nullable String subtitle) {
		this.subtitle = subtitle;
	}

	@NotNull
	public List<Author> getAuthors() {
		return BookAuthorTable.INSTANCE.searchBook(isbn);
	}

	@NotNull
	public String getPublisher() {
		return publisher;
	}

	public void setPublisher(@NotNull String publisher) {
		this.publisher = publisher;
	}

	@NotNull
	public Format getFormat() {
		return format;
	}

	public void setFormat(@NotNull Format format) {
		this.format = format;
	}
	//</editor-fold>

	//<editor-fold desc="Object Functions" defaultstate="collapsed">
	@Override
	public int hashCode() {
		int result = isbn.hashCode();
		result = 31 * result + title.hashCode();
		result = 31 * result + (subtitle != null ? subtitle.hashCode() : 0);
		result = 31 * result + publisher.hashCode();
		result = 31 * result + format.hashCode();
		return result;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Book)) return false;

		Book book = (Book) o;

		if (!isbn.equals(book.isbn)) return false;
		if (!title.equals(book.title)) return false;
		if (!Objects.equals(subtitle, book.subtitle)) return false;
		if (!publisher.equals(book.publisher)) return false;
		return format == book.format;
	}

	@Override
	public String toString() {
		return "Book{" +
				"isbn=" + isbn +
				", title='" + title + '\'' +
				", subtitle='" + subtitle + '\'' +
				", publisher='" + publisher + '\'' +
				", format=" + format +
				'}';
	}
	//</editor-fold>
}

File: src/main/kotlin/macro/library/book/Book.kt
package macro.library.book

import macro.library.ISendable
import macro.library.database.BookTable
import macro.library.external.GoogleBooks
import macro.library.external.OpenLibrary
import org.apache.logging.log4j.LogManager

/**
 * Created by Macro303 on 2019-Oct-30
 */
data class Book(
	val isbn: Isbn,
	var title: String,
	var subtitle: String?,
	var publisher: String?,
	var format: Format = Format.PAPERBACK,
	var openLibraryId: String? = null,
	var googleBooksId: String? = null,
	var goodreadsId: String? = null
) : Comparable<Book>, ISendable {

	fun add(): Book {
		BookTable.insert(this)
		return this
	}

	fun push(): Book {
		BookTable.update(this)
		return this
	}

	fun remove() {
		BookTable.delete(isbn)
	}

	override fun compareTo(other: Book): Int = comparator.compare(this, other)

	override fun toJson(full: Boolean): Map<String, Any?> {
		val output = mutableMapOf<String, Any?>(
			"ISBN" to isbn.toString(),
			"Title" to title,
			"Subtitle" to subtitle,
			"Format" to format.getDisplay(),
			"Images" to mutableMapOf<String, Any?>(
				"Small" to (if (openLibraryId != null) "https://covers.openlibrary.org/b/OLID/$openLibraryId-S.jpg" else if (googleBooksId != null) "http://books.google.com/books/content?id=$googleBooksId&printsec=frontcover&img=1&zoom=1" else null),
				"Medium" to (if (openLibraryId != null) "https://covers.openlibrary.org/b/OLID/$openLibraryId-M.jpg" else if (googleBooksId != null) "http://books.google.com/books/content?id=$googleBooksId&printsec=frontcover&img=1&zoom=5" else null),
				"Large" to (if (openLibraryId != null) "https://covers.openlibrary.org/b/OLID/$openLibraryId-L.jpg" else if (googleBooksId != null) "http://books.google.com/books/content?id=$googleBooksId&printsec=frontcover&img=1&zoom=10" else null)
			).toSortedMap()
		)
		if (full) {
			output["Publisher"] = publisher
			output["External IDs"] = mutableMapOf<String, Any?>(
				"Open Library" to openLibraryId,
				"Google Books" to googleBooksId,
				"Goodreads" to goodreadsId
			).toSortedMap()
		}
		return output.toSortedMap()
	}

	companion object {
		private val LOGGER = LogManager.getLogger(Book::class.java)
		private val comparator = compareBy(String.CASE_INSENSITIVE_ORDER, Book::title)
			.thenBy(nullsLast(), Book::subtitle)
			.thenBy(Book::format)
			.thenBy(Book::isbn)

		fun create(isbn: Isbn, format: Format = Format.PAPERBACK): Book? {
			var book = BookTable.selectUnique(isbn)
			if (book == null) {
				LOGGER.info("Book not found in DB")
				book = OpenLibrary.searchBook(isbn)
				if (book == null) {
					LOGGER.info("Book not found in Open Library")
					book = GoogleBooks.searchBook(isbn)
					if (book == null) {
						LOGGER.info("Book not found in Google Books")
						return null
					}
				} else
					book = GoogleBooks.selectBook(book)
			} else {
				book = OpenLibrary.selectBook(book)
				book = GoogleBooks.selectBook(book)
			}
			book.format = format
			return book.push()
		}
	}
}